src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        storage = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= storage.length) {
+            grow();
+        }
+        storage[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (size >= storage.length) {
+            grow();
+        }
+        moveAdd(index);
+        storage[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
-        return null;
+    public T remove(int index) {
+        checkIndex(index);
+        Object[] replaceStorage = new Object[storage.length];
+        System.arraycopy(storage, 0, replaceStorage, 0, index);
+        System.arraycopy(storage, index + 1, replaceStorage, index, (storage.length - index - 1));
+        T oldValue = storage[index];
+        updateStorage(replaceStorage);
```

Comment:
try to make one method for resizing the array, cause it is used in many methods, in order not to duplicate code